### PR TITLE
ENH: Add warm start to sparse tikhonov solvers

### PR DESCRIPTION
**Pull request by SalimBenchelabi**

### CODE REVIEW

File: iterreg/sparse/solvers.py [hunk 1]
```diff
@@ -167,15 +167,18 @@ def cd_primal_dual(X, y, prox=shrink, max_iter=100, f_store=1, verbose=False):
 
 @njit
 def cd(X, y, alpha, prox=shrink, pen=ell1, max_iter=1_000,
-       f_store=1, verbose=False):
+       f_store=1, verbose=False, w_0=None):
```
Comment (mathurinm) [inline iterreg/sparse/solvers.py]: We like to keep verbose as the last parameter, can you invert the last two ?

File: iterreg/sparse/solvers.py [hunk 2]
```diff
@@ -167,15 +167,18 @@ def cd_primal_dual(X, y, prox=shrink, max_iter=100, f_store=1, verbose=False):
...
     """Coordinate descent for the Tikhonov problem."""
     p = X.shape[1]
     X = np.asfortranarray(X)
     lc = np.zeros(p)
     for j in range(p):
         lc[j] = norm(X[:, j]) ** 2
     R = y.copy().astype(np.float64)
-    w = np.zeros(p)
+    if w_0 is None:
```
Comment (mathurinm) [inline iterreg/sparse/solvers.py]: changing w's type from None to a np.array does not break numba ?

File: iterreg/sparse/solvers.py [hunk 3]
```diff
@@ -221,11 +227,14 @@ def ista(X, y, alpha, prox=shrink, pen=ell1, max_iter=1_000, f_store=1,
...
-          verbose=False):
+          verbose=False, w_0=None):
     """Accelerated proximal gradient descent for the Tikhonov problem."""
     p = X.shape[1]
     L = norm(X, ord=2) ** 2
-    w = np.zeros(p)
+    if w_0 is None:
+        w = np.zeros(p)
+    else:
+        w = w_0
```
Comment (mathurinm) [inline iterreg/sparse/solvers.py]: beware that if w0 is defined outside, any modification to w here will affect w0 to in a side effect. It's safer to use `w = w0.copy()` IMO

File: iterreg/sparse/solvers.py (hunk 2 above)
Comment (SalimBenchelabi) [reply iterreg/sparse/solvers.py]: It seems to work, I had no error message

File: iterreg/sparse/solvers.py [hunk 4]
```diff
@@ -220,13 +229,16 @@ def ista(X, y, alpha, prox=shrink, pen=ell1, max_iter=1_000, f_store=1,
...
     """Accelerated proximal gradient descent for the Tikhonov problem."""
     p = X.shape[1]
     L = norm(X, ord=2) ** 2
-    w = np.zeros(p)
-    z = np.zeros(p)
+    if w_init is None:
+        w = np.zeros(p)
+    else:
+        w = w_init.copy()
+    z = w.copy()
```
Comment (mathurinm) [inline iterreg/sparse/solvers.py]: I think this opens a new (interesting !) question and should be implemented in another PR. if you revert this for me we can merge

File: iterreg/sparse/test_ell1.py [hunk 5]
```diff
@@ -45,3 +45,22 @@ def test_cd_ista_fista():
...
     w, _, _ = fista(X, y, alpha, max_iter=1_000)
     np.testing.assert_allclose(w, clf.coef_, atol=5e-4)
+
+
+def test_cd_warm_start():
+    X, y, _ = make_correlated_data(30, 50, random_state=12)
+    alpha = np.max(np.abs(X.T @ y)) / 100
+
+    # same to do 20 iter, or 10 iter, and 10 iter again with warm start:
+    w, _, E = cd(X, y, alpha, max_iter=20, f_store=1)
```
Comment (mathurinm) [inline iterreg/sparse/test_ell1.py]: Can you do: `for algo in [cd, ista]` since the code is the same ?
